crates/solana-listener/src/component/signature_realtime_scanner.rs

```diff
@@ -34,21 +38,34 @@ pub(crate) async fn process_realtime_logs(
             solana_client::nonblocking::pubsub_client::PubsubClient::new(config.solana_ws.as_str())
                 .await?;
 
-        // todo we have to also subscribe to the gas service events (new websocket stream running in
-        // parallel).
-        // Reason for parallel stream: https://solana.com/docs/rpc/websocket/logssubscribe :
-        // "The mentions field currently only supports one Pubkey string per method call. Listing
+        // Reason for subscribing to gateway and gas service separately
+        // (https://solana.com/docs/rpc/websocket/logssubscribe):
+        //
+        // "The `mentions` field currently only supports one Pubkey string per method call. Listing
         // additional addresses will result in an error."
-
-        let (ws_stream, _unsubscribe) = client
+        let (gateway_ws_stream, _unsubscribe) = client
             .logs_subscribe(
                 RpcTransactionLogsFilter::Mentions(vec![gateway_program_address.to_string()]),
                 RpcTransactionLogsConfig {
                     commitment: Some(CommitmentConfig::finalized()),
                 },
             )
             .await?;
-        let mut ws_stream = ws_stream.fuse();
+
+        let (gas_service_ws_stream, _unsubscribe) = client
+            .logs_subscribe(
+                RpcTransactionLogsFilter::Mentions(vec![gas_service_address.to_string()]),
+                RpcTransactionLogsConfig {
+                    commitment: Some(CommitmentConfig::finalized()),
+                },
+            )
+            .await?;
+
+        // We could use `futures::stream::select` here, but it only completes when both streams
+        // complete. As we want to complete when either stream completes, we need a custom
+        // combinator.
+        let mut ws_stream =
+            CompleteOnAny::new(gateway_ws_stream.fuse(), gas_service_ws_stream.fuse());
```

Comment:
I'll double check if we can use something off-the-shelf or something that wouldn't require us to bring out proc macros 🙈

Reply:
It's not very difficult to drop the proc-macro, but then the amount of code (almost-)duplication is quite big and not pretty either. So yeah, if we can find something self-contained would be best indeed.